mono/utils/mono-threads-api.h

```diff
@@ -22,35 +22,63 @@ MONO_BEGIN_DECLS
 This API is experimental. It will eventually be required to properly use the rest of the raw-omp embedding API.
 */
 
+typedef struct _MonoStackData {
+	const char *function_name;
+} MonoStackData;
```

Comment:
The issue I have with this approach is that by casting `gpointer*` to `MonoStackData*` (in all the `mono_threads_{enter,exit}_gc_{safe,unsafe}_region*` functions you are encoding in the API - without any other check than hope - that you are not going to change this `MonoStackData` structure to anything else than a `char*` (or anything equivalent to a `gpointer`). And if you do happen to change that structure for some reasons, and you forget to do the same on XI/XM/XA side of things, then it's going to break with random stack corruption. So how will you protect against that?

Reply:
Please makes this an opaque type with getters/setters, and add some verification logic in the cast.

Reply:
@luhenry I thought about this a bit. I think if the struct evolves, the _internal function would need flags or a size, or the struct would have to start with flags (or bitflag) or a size. Or in fact the struct is not great and today we can/should pass a separate const char* parameter. Structs are preferred because intermediate functions can just pass them through by pointer without worrying about their precise contents -- evolving the struct takes work, but less work.

Basically a flag per field, to indicate it is present, or you can use size. Size is easier to initialize, a little awkward to use. flags or bitfield is awkward to initialize but easy to check. Flags or bitfield get used up faster but probably slow enough. It can be a flag per field, easy to understand but used up fast or a "version". 

@alexanderkyte no verification logic is particlarly possible on the cast due to the v1 API, but getter/setter is. Well, we could pass in a size.

Also, to be clear, there is an extra subtletly here. "Normally" the v1 would likely be adapted to v2 with more of a copy in/out. The reason for the cast is because the pointer doubles as a stack pointer. I didn't consider that it could kinda be replaced in the deeper stack. I think that was a slight error though.

However, upon further thought:

How about more like this:

struct MonoStackData {
  union {
    gint32 as_int; // this always present
    struct {
     gbool has_stack_ptr : 1;  // required, but still a flag
     gbool has_function_name : 1; // optional 
   };
  };
  gpointer stack_ptr;
  const char *function_name;
};

Reply:
To be clear, I agree. What I have in the PR is a bit short sighted and I'll try something more evolvable and safe instead. Even those flags aren't the full story for API evolution -- you need in flags_required, in flags_optional, and out flags_honored, but likely people lose then lose their taste for compatible evolution. Just one set of flags is all people can stomach, and hand wave around the larger problem.